In [1]:
#Data Manipulation
import pandas as pd
import numpy as np

#Data Visualization
pd.plotting.register_matplotlib_converters()
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns

In [2]:
df_norm = pd.read_csv("healthcare-dataset-stroke-data.csv", index_col="id")
df_synt = pd.read_csv("synthetic_train.csv", index_col="id")

pd.set_option('display.max_columns', None)

print(f"Normal dataset unique columns: {list(df_norm.columns)}")
print(f"Normal dataset number of rows and columns: {df_norm.shape}")
print("----------------------------------------------------------------------------")
print(f"Synthetic dataset unique columns: {list(df_synt.columns)}")
print(f"Synthetic dataset number of rows and columns: {df_synt.shape}")

Normal dataset unique columns: ['gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'work_type', 'Residence_type', 'avg_glucose_level', 'bmi', 'smoking_status', 'stroke']
Normal dataset number of rows and columns: (5110, 11)
----------------------------------------------------------------------------
Synthetic dataset unique columns: ['gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'work_type', 'Residence_type', 'avg_glucose_level', 'bmi', 'smoking_status', 'stroke']
Synthetic dataset number of rows and columns: (15304, 11)


In [3]:
df_total = pd.concat([df_norm, df_synt])

In [4]:
print(df_total.shape)
df_total

(20414, 11)


gender   age  hypertension  heart_disease ever_married      work_type  \
id                                                                             
9046     Male  67.0             0              1          Yes        Private   
51676  Female  61.0             0              0          Yes  Self-employed   
31112    Male  80.0             0              1          Yes        Private   
60182  Female  49.0             0              0          Yes        Private   
1665   Female  79.0             1              0          Yes  Self-employed   
...       ...   ...           ...            ...          ...            ...   
15299  Female  22.0             0              0           No       Govt_job   
15300  Female  46.0             1              0          Yes        Private   
15301  Female  75.0             0              0          Yes  Self-employed   
15302    Male  46.0             0              0          Yes        Private   
15303  Female  14.0             0              0           No        Private   

      Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
id                                                                      
9046           Urban             228.69  36.6  formerly smoked       1  
51676          Rural             202.21   NaN     never smoked       1  
31112          Rural             105.92  32.5     never smoked       1  
60182          Urban             171.23  34.4           smokes       1  
1665           Rural             174.12  24.0     never smoked       1  
...              ...                ...   ...              ...     ...  
15299          Urban              72.63  19.5     never smoked       0  
15300          Urban             101.19  32.1     never smoked       0  
15301          Urban              87.69  26.2     never smoked       0  
15302          Rural             101.13  22.5          Unknown       0  
15303          Rural              85.12  24.7     never smoked       0  

[20414 rows x 11 columns]

In [5]:
df_total.isnull().sum()

gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

In [6]:
df_total = df_total.dropna()

In [7]:
df = df_total.copy()
df

gender   age  hypertension  heart_disease ever_married      work_type  \
id                                                                             
9046     Male  67.0             0              1          Yes        Private   
31112    Male  80.0             0              1          Yes        Private   
60182  Female  49.0             0              0          Yes        Private   
1665   Female  79.0             1              0          Yes  Self-employed   
56669    Male  81.0             0              0          Yes        Private   
...       ...   ...           ...            ...          ...            ...   
15299  Female  22.0             0              0           No       Govt_job   
15300  Female  46.0             1              0          Yes        Private   
15301  Female  75.0             0              0          Yes  Self-employed   
15302    Male  46.0             0              0          Yes        Private   
15303  Female  14.0             0              0           No        Private   

      Residence_type  avg_glucose_level   bmi   smoking_status  stroke  
id                                                                      
9046           Urban             228.69  36.6  formerly smoked       1  
31112          Rural             105.92  32.5     never smoked       1  
60182          Urban             171.23  34.4           smokes       1  
1665           Rural             174.12  24.0     never smoked       1  
56669          Urban             186.21  29.0  formerly smoked       1  
...              ...                ...   ...              ...     ...  
15299          Urban              72.63  19.5     never smoked       0  
15300          Urban             101.19  32.1     never smoked       0  
15301          Urban              87.69  26.2     never smoked       0  
15302          Rural             101.13  22.5          Unknown       0  
15303          Rural              85.12  24.7     never smoked       0  

[20213 rows x 11 columns]

In [8]:
#Label Encoding

df.loc[df.gender == "Male", "gender"] = 0
df.loc[df.gender != 0, "gender"] = 1
df["gender"] = df["gender"].astype(int)

df.loc[df.ever_married == "Yes", "ever_married"] = 0
df.loc[df.ever_married != 0, "ever_married"] = 1
df["ever_married"] = df["ever_married"].astype(int)

df.loc[df.Residence_type == "Urban", "Residence_type"] = 0
df.loc[df.Residence_type != 0, "Residence_type"] = 1
df["Residence_type"] = df["Residence_type"].astype(int)

# Standartization

df["avg_glucose_level"] = (df.avg_glucose_level - df.avg_glucose_level.mean()) / df.avg_glucose_level.std()
df["age"] = (df.age - df.age.mean()) / df.age.std()
df["bmi"] = (df.bmi - df.bmi.mean()) / df.bmi.std()

In [9]:
df = pd.get_dummies(data=df, columns=["work_type", "smoking_status"])

In [10]:
print(df.shape)
df

(20213, 18)


gender       age  hypertension  heart_disease  ever_married  \
id                                                                   
9046        0  1.161213             0              1             0   
31112       0  1.759523             0              1             0   
60182       1  0.332785             0              0             0   
1665        1  1.713499             1              0             0   
56669       0  1.805547             0              0             0   
...       ...       ...           ...            ...           ...   
15299       1 -0.909858             0              0             1   
15300       1  0.194713             1              0             0   
15301       1  1.529404             0              0             0   
15302       0  0.194713             0              0             0   
15303       1 -1.278049             0              0             1   

       Residence_type  avg_glucose_level       bmi  stroke  \
id                                                           
9046                0           4.250449  1.181714       1   
31112               1           0.404996  0.597816       1   
60182               0           2.450663  0.868403       1   
1665                1           2.541185 -0.612706       1   
56669               0           2.919873  0.099365       1   
...               ...                ...       ...     ...   
15299               0          -0.637728 -1.253571       0   
15300               0           0.256841  0.540850       0   
15301               0          -0.166012 -0.299395       0   
15302               1           0.254961 -0.826328       0   
15303               1          -0.246511 -0.513016       0   

       work_type_Govt_job  work_type_Never_worked  work_type_Private  \
id                                                                     
9046                    0                       0                  1   
31112                   0                       0                  1   
60182                   0                       0                  1   
1665                    0                       0                  0   
56669                   0                       0                  1   
...                   ...                     ...                ...   
15299                   1                       0                  0   
15300                   0                       0                  1   
15301                   0                       0                  0   
15302                   0                       0                  1   
15303                   0                       0                  1   

       work_type_Self-employed  work_type_children  smoking_status_Unknown  \
id                                                                           
9046                         0                   0                       0   
31112                        0                   0                       0   
60182                        0                   0                       0   
1665                         1                   0                       0   
56669                        0                   0                       0   
...                        ...                 ...                     ...   
15299                        0                   0                       0   
15300                        0                   0                       0   
15301                        1                   0                       0   
15302                        0                   0                       1   
15303                        0                   0                       0   

       smoking_status_formerly smoked  smoking_status_never smoked  \
id                                                                   
9046                                1                            0   
31112                               0                            1   
60182                               0                    

In [11]:
features = list(df.columns)
features.remove("stroke")
print(features)

['gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'Residence_type', 'avg_glucose_level', 'bmi', 'work_type_Govt_job', 'work_type_Never_worked', 'work_type_Private', 'work_type_Self-employed', 'work_type_children', 'smoking_status_Unknown', 'smoking_status_formerly smoked', 'smoking_status_never smoked', 'smoking_status_smokes']


In [12]:
X = df.copy()
y = X.pop("stroke")

#print(X.head(1))
print(X.dtypes)

gender                              int32
age                               float64
hypertension                        int64
heart_disease                       int64
ever_married                        int32
Residence_type                      int32
avg_glucose_level                 float64
bmi                               float64
work_type_Govt_job                  uint8
work_type_Never_worked              uint8
work_type_Private                   uint8
work_type_Self-employed             uint8
work_type_children                  uint8
smoking_status_Unknown              uint8
smoking_status_formerly smoked      uint8
smoking_status_never smoked         uint8
smoking_status_smokes               uint8
dtype: object


In [13]:
auc = []

In [50]:
import sklearn
from sklearn.svm import SVR
from sklearn.model_selection import cross_val_score

#print(sklearn.metrics.get_scorer_names())

"""
for n in ['linear', 'poly', 'rbf', 'sigmoid']:
    SVRmodel = SVR(gamma='auto', kernel=n)
    SVRmodel.fit(X, y)
    scores = cross_val_score(SVRmodel, X, y,
                              cv=5,
                              scoring='roc_auc')
    
    print("Used: ", n, " kernel")
    auc_score = sum(scores) / len(scores)
    auc.append(auc_score)
    print("Avarage AUC score:", auc_score)
"""

SVRmodel = SVR(gamma='auto', kernel="rbf")
SVRmodel.fit(X, y)

SVR(gamma='auto')

In [23]:
auc

[0.6615539079200584,
 0.6615539079200584,
 0.6630333831060288,
 0.671828563797656,
 0.6739580589588217,
 0.5992652481237254,
 0.6822645770075757,
 0.6618861329056387,
 0.44222623409677697,
 0.5992652481237254,
 0.5992652481237254,
 0.6822645770075757,
 0.6618861329056387]

Used:  linear  kernel
Avarage AUC score: 0.5992652481237254
Used:  poly  kernel
Avarage AUC score: 0.6822645770075757
Used:  rbf  kernel
Avarage AUC score: 0.6618861329056387
Used:  sigmoid  kernel
Avarage AUC score: 0.44222623409677697

In [25]:
df_test = pd.read_csv("test.csv", index_col="id")
df_test

gender   age  hypertension  heart_disease ever_married work_type  \
id                                                                        
15304  Female  57.0             0              0          Yes   Private   
15305    Male  70.0             1              0          Yes   Private   
15306  Female   5.0             0              0           No  children   
15307  Female  56.0             0              0          Yes  Govt_job   
15308    Male  32.0             0              0          Yes   Private   
...       ...   ...           ...            ...          ...       ...   
25503  Female  27.0             0              0           No   Private   
25504    Male  49.0             0              0          Yes   Private   
25505  Female   3.0             0              0           No  children   
25506    Male  31.0             0              0          Yes   Private   
25507  Female   2.0             0              0           No  children   

      Residence_type  avg_glucose_level   bmi smoking_status  
id                                                            
15304          Rural              82.54  33.4        Unknown  
15305          Urban              72.06  28.5        Unknown  
15306          Urban             103.72  19.5        Unknown  
15307          Urban              69.24  41.4         smokes  
15308          Rural             111.15  30.1         smokes  
...              ...                ...   ...            ...  
25503          Urban              75.77  17.6   never smoked  
25504          Urban             102.91  26.7        Unknown  
25505          Rural             104.04  18.3        Unknown  
25506          Urban              82.41  28.7   never smoked  
25507          Urban              85.12  14.8        Unknown  

[10204 rows x 10 columns]

In [26]:
#Label Encoding

df_test.loc[df_test.gender == "Male", "gender"] = 0
df_test.loc[df_test.gender != 0, "gender"] = 1
df_test["gender"] = df_test["gender"].astype(int)

df_test.loc[df_test.ever_married == "Yes", "ever_married"] = 0
df_test.loc[df_test.ever_married != 0, "ever_married"] = 1
df_test["ever_married"] = df_test["ever_married"].astype(int)

df_test.loc[df_test.Residence_type == "Urban", "Residence_type"] = 0
df_test.loc[df_test.Residence_type != 0, "Residence_type"] = 1
df_test["Residence_type"] = df_test["Residence_type"].astype(int)

# Standartization

df_test["avg_glucose_level"] = (df_test.avg_glucose_level - df_test.avg_glucose_level.mean()) / df_test.avg_glucose_level.std()
df_test["age"] = (df_test.age - df_test.age.mean()) / df_test.age.std()
df_test["bmi"] = (df_test.bmi - df_test.bmi.mean()) / df_test.bmi.std()

In [27]:
df_test = pd.get_dummies(data=df_test, columns=["work_type", "smoking_status"])

In [28]:
df_test

gender       age  hypertension  heart_disease  ever_married  \
id                                                                   
15304       1  0.714727             0              0             0   
15305       0  1.330901             1              0             0   
15306       1 -1.749966             0              0             1   
15307       1  0.667329             0              0             0   
15308       0 -0.470222             0              0             0   
...       ...       ...           ...            ...           ...   
25503       1 -0.707211             0              0             1   
25504       0  0.335544             0              0             0   
25505       1 -1.844762             0              0             1   
25506       0 -0.517619             0              0             0   
25507       1 -1.892160             0              0             1   

       Residence_type  avg_glucose_level       bmi  work_type_Govt_job  \
id                                                                       
15304               1          -0.256914  0.765927                   0   
15305               0          -0.661323  0.038262                   0   
15306               0           0.560393 -1.298267                   0   
15307               0          -0.770143  1.953953                   1   
15308               1           0.847107  0.275867                   0   
...               ...                ...       ...                 ...   
25503               0          -0.518159 -1.580423                   0   
25504               0           0.529136 -0.229044                   0   
25505               1           0.572742 -1.476471                   0   
25506               0          -0.261931  0.067962                   0   
25507               0          -0.157355 -1.996232                   0   

       work_type_Never_worked  work_type_Private  work_type_Self-employed  \
id                                                                          
15304                       0                  1                        0   
15305                       0                  1                        0   
15306                       0                  0                        0   
15307                       0                  0                        0   
15308                       0                  1                        0   
...                       ...                ...                      ...   
25503                       0                  1                        0   
25504                       0                  1                        0   
25505                       0                  0                        0   
25506                       0                  1                        0   
25507                       0                  0                        0   

       work_type_children  smoking_status_Unknown  \
id                                                  
15304                   0                       1   
15305                   0                       1   
15306                   1                       1   
15307                   0                       0   
15308                   0                       0   
...                   ...                     ...   
25503                   0                       0   
25504                   0                       1   
25505                   1                       1   
25506                   0                       0   
25507                   1                       1   

       smoking_status_formerly smoked  smoking_status_never smoked  \
id                                                                   
15304                               0                            0   
15305                               0                            0   
15306                               0                            0   
15307                               0                            0   
15308                

In [51]:
predictions = SVRmodel.predict(df_test)
#for pred in predictions:
    #print(pred)
    
print("----------------------------------------------")
normalized_predictions = (predictions-min(predictions))/(max(predictions)-min(predictions))    
for norm_pred in normalized_predictions:
    print(norm_pred)
    
outputs = normalized_predictions

----------------------------------------------
0.38017137985146104
0.38479152637208586
0.3631005894518736
0.37504178409666195
0.37274644260770323
0.37992561055548274
0.3757128118729843
0.39325376771309933
0.3701224921373566
0.38116724412258546
0.3790046055499823
0.40105927911900646
0.361319265677719
0.3809526035658161
0.367098866348624
0.38232490197147395
0.38250469655598507
0.3737526420206333
0.3775861458999231
0.381495716454291
0.3733122838378918
0.38074652734611686
0.3590540147639177
0.3785970996726344
0.37984855165189624
0.38285275974466315
0.37540077839025715
0.38027814456153086
0.3815451830592109
0.3777893257099061
0.3796386308748516
0.38088415903594125
0.37863054178681077
0.3820312862174947
0.3802351902418593
0.35366943931993094
0.37877626029322287
0.36904030965362034
0.35461538059422776
0.3809185632293833
0.3784566973225986
0.37884807584539576
0.35936508925002336
0.3793387091768993
0.36910919687583044
0.3643210032432793
0.380333140665413
0.3215496508890521
0.37627955515527867
0

In [54]:
sub_file = open("subSVR.csv", "w")

ID = 15304
run = 0

sub_file.write("id,stroke\n")
for output in outputs:
    sub_file.write(f"{ID},{output}\n")
    ID += 1
    run +=1
    
print(run)

10204
